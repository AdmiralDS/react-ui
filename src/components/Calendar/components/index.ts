export * from './Years';
export * from './Button';
export * from './Day';
export * from './DayNames';
export * from './Month';
export * from './Months';
export * from './Panel';
export * from './Week';
