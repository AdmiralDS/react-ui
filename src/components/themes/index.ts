export * from './light';
export * from './dark';
export type { Theme } from './common';
