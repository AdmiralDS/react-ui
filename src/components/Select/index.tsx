import { ComponentDimension } from '#src/components/input';
import type { InputStatus } from '#src/components/input';
import { StatusIcon } from '#src/components/input/StatusIcon';
import { OpenStatusButton } from '#src/components/OpenStatusButton';
import type { HTMLAttributes, KeyboardEvent, ReactElement, ReactNode } from 'react';
import * as React from 'react';
import styled, { css, DefaultTheme, FlattenInterpolation, ThemeProps } from 'styled-components';
import { keyboardKey } from '#src/components/common/keyboardKey';
import { refSetter } from '#src/components/common/utils/refSetter';
import { Dropdown } from '#src/components/Dropdown';
import { TYPOGRAPHY } from '#src/components/Typography';

export const SELECT_DIMENSIONS = ['xl', 'm', 's'] as const;
export type SelectDimension = typeof SELECT_DIMENSIONS[number];

const heights = css<{ dimension: SelectDimension }>`
  height: ${({ dimension }) => {
    switch (dimension) {
      case 'xl':
        return '56px';
      case 'm':
        return '40px';
      case 's':
        return '32px';
      default:
        return '56px';
    }
  }};
`;

const styleText = css<{ dimension: SelectDimension }>`
  font-size: ${({ dimension }) => (dimension === 's' ? 14 : 16)}px;
  line-height: ${({ dimension }) => (dimension === 's' ? 20 : 24)}px;
  font-family: ${TYPOGRAPHY.fontFamily};
`;

const SelectComponent = styled.div<{ dimension: SelectDimension; disabled?: boolean }>`
  position: relative;
  padding: 0 ${({ dimension }) => (dimension === 's' ? 12 : 16)}px;
  color: ${({ theme, disabled }) => (disabled ? theme.color.text.tertiary : theme.color.text.primary)};
  display: flex;
  user-select: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  border: 0;
  align-items: center;
  justify-content: space-between;
  ${styleText};
  box-sizing: border-box;
  border-radius: 4px;
  height: 100%;

  [data-status='error'] & {
    border-color: ${(props) => props.theme.color.status.danger};
  }

  [data-read-only] & {
    user-select: none;
    background-color: ${(props) => props.theme.color.background.tertiary};
    border-color: ${(props) => props.theme.color.background.tertiary};
  }
`;

const Placeholder = styled.div<{ dimension: SelectDimension; focused: boolean; disabled?: boolean }>`
  color: ${({ theme, focused, disabled }) => {
    if (focused) return theme.color.basic.disable;
    if (disabled) return theme.color.text.tertiary;
    return theme.color.text.secondary;
  }};
  pointer-events: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  ${styleText}
`;

const StyledDropDown = styled(Dropdown)<{
  alignDropdown?: string;
  dropMaxHeight: string | number;
  cssMixin?: FlattenInterpolation<ThemeProps<DefaultTheme>>;
}>`
  padding: 8px 0;
  overflow-x: hidden;
  overflow-y: auto;
  ${(p) => (p.alignDropdown ? `align-self: ${p.alignDropdown}` : '')};
  max-height: ${(p) => p.dropMaxHeight};
  min-width: 100%;
  ${(p) => p.cssMixin || ''}
`;

const IconPanel = styled.div<{ disabled?: boolean; dimension?: ComponentDimension }>`
  flex: 0 0 auto;

  display: flex;
  align-items: center;

  margin-left: 10px;

  & > * {
    display: block;
    width: ${({ dimension }) => (dimension === 's' ? 20 : 24)}px;
    height: ${({ dimension }) => (dimension === 's' ? 20 : 24)}px;
  }

  & > *:not(:first-child) {
    margin-left: 8px;
  }
`;

const SelectWrapper = styled.div<{
  disabled?: boolean;
  dimension: SelectDimension;
  focused: boolean;
}>`
  box-sizing: border-box;

  cursor: pointer;

  &&[disabled],
  &&[data-read-only] {
    cursor: default;
  }

  background: ${({ theme, disabled }) => (disabled ? theme.color.background.tertiary : theme.color.background.primary)};
  position: relative;
  border-radius: 4px;
  ${heights};

  border-color: ${({ theme, disabled, focused }) => {
    if (disabled) return 'transparent';
    if (focused) return theme.color.basic.press;
    return theme.color.basic.tertiary;
  }};

  [data-focus-within] &,
  &:focus {
    border-color: ${(p) => p.theme.color.basic.press};
  }

  border-width: 1px;
  border-style: solid;

  &:hover {
    border-color: ${({ theme, disabled }) => (disabled ? 'transparent' : theme.color.basic.hover)};
  }

  &:focus {
    outline: none;
  }

  &:active {
    outline: none;
  }

  &[data-status='error'] {
    border-color: ${(p) => p.theme.color.status.danger};
  }

  &[data-status='success'] {
    border-color: ${(props) => props.theme.color.status.success};
  }
`;

const WrapperChild = styled.div`
  flex: 1 1 auto;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export interface SelectProps extends Omit<HTMLAttributes<HTMLDivElement>, 'onChange'> {
  /** Отключение селекта */
  disabled?: boolean;
  /** Размер Меню */
  dimension?: SelectDimension;
  /** Значение селекта */
  value: number | string | null;
  /** Функция, которая возвращает реакт-компонент с контентом для отображения в cелекте */
  renderInputValue?: (value: number | string | null) => ReactNode;
  /** Обработчик для изменения состояния селекта */
  onChange: (value: number | string | null) => void;
  /** Колбек на открытие селекта */
  onOpen?: () => void;
  /** Колбек на закрытие селекта */
  onClose?: () => void;
  /** Референс на контейнер для правильного позиционирования выпадающего списка */
  portalTargetRef?: React.RefObject<HTMLElement>;

  /**
   * Позволяет выравнивать позицию дропдаун контейнера относительно селекта.
   * Принимает стандартные значения css свойства align-self (auto | flex-start | flex-end | center | baseline | stretch)
   */
  alignDropdown?: 'auto' | 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch';

  /**
   *  Задает максимальную высоту контейнера с опциями
   */
  dropMaxHeight?: string | number;

  /** Установка статуса поля */
  status?: InputStatus;

  /** Отображать иконку статуса */
  displayStatusIcon?: boolean;

  /** Иконки для отображения в правом углу поля */
  icons?: React.ReactNode;

  /** Отображает значение без возможности изменения */
  readOnly?: boolean | undefined;

  /** Позволяет добавлять миксин на дроп контейнер созданный с помощью styled css  */
  dropContainerCssMixin?: FlattenInterpolation<ThemeProps<DefaultTheme>>;
}

export const Select = React.forwardRef<HTMLDivElement, SelectProps>(
  (
    {
      portalTargetRef,
      children,
      renderInputValue = (value) => value,
      className,
      disabled,
      dimension = 'm',
      value,
      onChange,
      onOpen,
      onClose,
      alignDropdown,
      dropMaxHeight = '300px',
      icons,
      status,
      displayStatusIcon,
      readOnly,
      dropContainerCssMixin,
      ...props
    },
    ref,
  ) => {
    const [open, setOpen] = React.useState(false);
    const [hovered, setHovered] = React.useState('');
    const refWrapper = React.useRef<HTMLInputElement>(null);
    const childrenArray = React.Children.toArray(children);
    const findOptionValue = (option: (React.ReactChild | React.ReactFragment | React.ReactPortal)[]) => {
      if (React.isValidElement(option[0]) && 'props' in option[0]) {
        return option[0].props.value;
      }
    };

    const hoverIndex = React.useMemo(
      () =>
        childrenArray?.findIndex((child) => {
          if (React.isValidElement(child) && 'props' in child) {
            return child.props.value === +hovered;
          }
          return -1;
        }),
      [childrenArray, hovered],
    );

    const findNextHoverValue = React.useCallback(() => {
      const nextAbledOptionValue = findOptionValue(childrenArray.slice(hoverIndex + 1));
      if (nextAbledOptionValue) return nextAbledOptionValue;
      return findOptionValue(childrenArray);
    }, [hoverIndex, childrenArray]);

    const findPrevHoverValue = React.useCallback(() => {
      const sliceInd = hoverIndex === -1 ? undefined : hoverIndex;
      const prevAbledOptionValue = findOptionValue(childrenArray.slice(0, sliceInd).reverse());
      if (prevAbledOptionValue) return prevAbledOptionValue;
      return findOptionValue(childrenArray.slice().reverse());
    }, [hoverIndex, childrenArray]);

    const handleKeyDownItem = (e: KeyboardEvent) => {
      const code = keyboardKey.getCode(e);
      e.preventDefault();
      switch (code) {
        case keyboardKey[' ']:
        case keyboardKey.Enter: {
          setOpen(false);
          onClose?.();
          onChange && onChange(hovered);
          setHovered(hovered);
          (refWrapper.current as HTMLElement).focus();
          break;
        }
        case keyboardKey.ArrowUp: {
          const prevValue = findPrevHoverValue();
          if (!prevValue) break;
          setHovered(prevValue);
          break;
        }
        case keyboardKey.ArrowDown: {
          const nextValue = findNextHoverValue();
          if (!nextValue) break;
          setHovered(nextValue);
          break;
        }
        case keyboardKey.Tab:
        case keyboardKey.Escape: {
          onClose?.();
          setOpen(false);
          (refWrapper.current as HTMLElement).focus();
          break;
        }
      }
    };

    const handleClickItem = (e: React.MouseEvent<HTMLElement>, disabled: boolean) => {
      const value = e?.currentTarget.getAttribute('value') ?? '';
      !disabled && onChange && onChange(value);
      refWrapper.current?.focus();
    };

    const renderChildrenDropDown = () => {
      return React.Children.map(children, (child: React.ReactNode) => {
        if (!React.isValidElement(child)) {
          return null;
        }

        const detectedSubGroup = child.props.children?.some?.((element: ReactElement) => Array.isArray(element));

        if (detectedSubGroup) {
          return child.props.children.flat().map((child: ReactElement) => {
            return React.cloneElement(child, {
              disabled: disabled || child?.props?.disabled,
              onClick: (e: React.MouseEvent<HTMLElement>) => handleClickItem(e, child.props.disabled),
              onKeyDown: handleKeyDownItem,
              dimension: dimension === 'xl' ? 'l' : dimension,
              selected: value?.toString() === child.props.value?.toString(),
              hovered: hovered === child.props.value,
              ...child.props,
            });
          });
        }

        return React.cloneElement(child, {
          disabled: disabled || child.props.disabled,
          onClick: (e: React.MouseEvent<HTMLElement>) => handleClickItem(e, child.props.disabled),
          onKeyDown: handleKeyDownItem,
          dimension: dimension === 'xl' ? 'l' : dimension,
          selected: value?.toString() === child.props.value?.toString(),
          hovered: hovered === child.props.value,
          ...child.props,
        });
      });
    };

    const clickOutside = (e: Event) => {
      if (e.target && refWrapper.current?.contains(e.target as Node)) {
        return;
      }
      setOpen(false);
      open && onClose?.();
    };

    const handleKeyDown = (e: KeyboardEvent<HTMLDivElement>) => {
      const code = keyboardKey.getCode(e);
      if (code === keyboardKey[' '] && !disabled && !open) {
        setOpen(true);
        e.preventDefault();
        !open ? onOpen?.() : onClose?.();
      }
      props.onKeyDown?.(e);
    };

    const onClick = (e: React.MouseEvent<HTMLDivElement>) => {
      if (!disabled) {
        setOpen((open) => !open);
        !open ? onOpen?.() : onClose?.();
      }
      props.onClick?.(e);
    };

    const renderSelectValue = () =>
      props.placeholder && !value ? (
        <Placeholder disabled={disabled} focused={open} dimension={dimension}>
          {props.placeholder}
        </Placeholder>
      ) : (
        <WrapperChild>{renderInputValue(value)}</WrapperChild>
      );

    const iconArray = React.Children.toArray(icons);

    if (displayStatusIcon) {
      iconArray.push(<StatusIcon key="status-icon" status={status} aria-hidden />);
    }

    if (!readOnly) {
      iconArray.push(
        <OpenStatusButton
          key="open-status-icon"
          $isOpen={open}
          data-disabled={disabled ? true : undefined}
          aria-hidden
        />,
      );
    }

    return (
      <SelectWrapper
        ref={refSetter(ref, refWrapper)}
        className={className}
        disabled={disabled}
        data-disabled={disabled ? true : undefined}
        dimension={dimension}
        role={readOnly ? undefined : 'button'}
        {...props}
        onKeyDown={handleKeyDown}
        tabIndex={disabled ? undefined : 0}
        onClick={readOnly ? () => null : onClick}
        data-status={status}
        data-read-only={readOnly ? true : undefined}
        aria-readonly={readOnly ? true : undefined}
        focused={open}
      >
        <SelectComponent dimension={dimension} disabled={disabled}>
          {renderSelectValue()}
          <IconPanel disabled={disabled} dimension={dimension}>
            {iconArray}
          </IconPanel>
        </SelectComponent>
        {open && (
          <StyledDropDown
            targetRef={portalTargetRef || refWrapper}
            data-dimension={dimension === 'xl' ? 'l' : dimension}
            onClickOutside={clickOutside}
            alignDropdown={alignDropdown}
            dropMaxHeight={dropMaxHeight}
            cssMixin={dropContainerCssMixin}
          >
            {renderChildrenDropDown()}
          </StyledDropDown>
        )}
      </SelectWrapper>
    );
  },
);

Select.displayName = 'Select';
