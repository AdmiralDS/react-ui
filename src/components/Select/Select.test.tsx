import { DropDownItem } from '#src/components/DropDownItem';
import { Select } from '#src/components/Select';
import { LIGHT_THEME } from '#src/components/themes';
import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { ThemeProvider } from 'styled-components';

describe('Select', () => {
  beforeEach(() => {
    jest.clearAllTimers();
  });

  it('should render component', () => {
    const wrapper = render(
      <ThemeProvider theme={LIGHT_THEME}>
        <Select value={'Селект'} onChange={() => ''}>
          <DropDownItem value={1}>Option</DropDownItem>
        </Select>
      </ThemeProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render small components', () => {
    const wrapper = render(
      <ThemeProvider theme={LIGHT_THEME}>
        <Select dimension={'s'} value={'Селект'} onChange={() => ''}>
          <DropDownItem value={'1'}>Option</DropDownItem>
        </Select>
      </ThemeProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should open dropDown when user clicks on component', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <Select data-testid="select" value={'Селект'} onChange={() => ''}>
          <DropDownItem data-testid="select-option" value={'1'}>
            Option one
          </DropDownItem>
          <DropDownItem data-testid="select-option" value={'2'}>
            Option two
          </DropDownItem>
          <DropDownItem data-testid="select-option" value={'3'}>
            Option three
          </DropDownItem>
        </Select>
      </ThemeProvider>,
    );

    fireEvent.click(screen.getByTestId('select'));

    expect(screen.getByText(`Option one`)).toHaveTextContent(`Option one`);
  });

  it('should work with clicks and touch events', () => {
    const handleChange = jest.fn();
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <Select value="empty" onChange={handleChange}>
          <DropDownItem value="option">Option</DropDownItem>
        </Select>
      </ThemeProvider>,
    );

    fireEvent.click(screen.getByText(`empty`));
    fireEvent.click(screen.getByText(`Option`));
    expect(handleChange).toHaveBeenCalledTimes(1);
  });
});
