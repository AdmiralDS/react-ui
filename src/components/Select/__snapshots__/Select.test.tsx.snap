// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c6 {
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c6 *[fill^='#'] {
  fill: #626F84;
}

.c6.c6[data-disabled] {
  pointer-events: none;
  fill: #A7B0BE;
}

.c6:not([data-disabled]):hover {
  cursor: pointer;
}

.c6:not([data-disabled]):hover *[fill^='#'] {
  fill: #0A54C2;
}

.c3 {
  position: relative;
  padding: 0 16px;
  color: #2B313B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  border: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 16px;
  line-height: 24px;
  font-family: 'VTB Group UI',sans-serif;
  box-sizing: border-box;
  border-radius: 4px;
  height: 100%;
}

[data-status='error'] .c2 {
  border-color: #DA0B20;
}

[data-read-only] .c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #E2E5E9;
  border-color: #E2E5E9;
}

.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 10px;
}

.c5 > * {
  display: block;
  width: 24px;
  height: 24px;
}

.c5 > *:not(:first-child) {
  margin-left: 8px;
}

.c1 {
  box-sizing: border-box;
  cursor: pointer;
  background: #FFFFFF;
  position: relative;
  border-radius: 4px;
  height: 40px;
  border-color: #8A96A8;
  border-width: 1px;
  border-style: solid;
}

.c1.c1[disabled],
.c1.c1[data-read-only] {
  cursor: default;
}

[data-focus-within] .c0,
.c1:focus {
  border-color: #0949AA;
}

.c1:hover {
  border-color: #0A54C2;
}

.c1:focus {
  outline: none;
}

.c1:active {
  outline: none;
}

.c1[data-status='error'] {
  border-color: #DA0B20;
}

.c1[data-status='success'] {
  border-color: #2ABB5B;
}

.c4 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<body>
    <div>
      <div
        class="c0 c1"
        role="button"
        tabindex="0"
      >
        <div
          class="c2 c3"
        >
          <div
            class="c4"
          >
            Селект
          </div>
          <div
            class="c5"
          >
            <div
              aria-hidden="true"
              class="c6"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": .c6 {
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c6 *[fill^='#'] {
  fill: #626F84;
}

.c6.c6[data-disabled] {
  pointer-events: none;
  fill: #A7B0BE;
}

.c6:not([data-disabled]):hover {
  cursor: pointer;
}

.c6:not([data-disabled]):hover *[fill^='#'] {
  fill: #0A54C2;
}

.c3 {
  position: relative;
  padding: 0 16px;
  color: #2B313B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  border: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 16px;
  line-height: 24px;
  font-family: 'VTB Group UI',sans-serif;
  box-sizing: border-box;
  border-radius: 4px;
  height: 100%;
}

[data-status='error'] .c2 {
  border-color: #DA0B20;
}

[data-read-only] .c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #E2E5E9;
  border-color: #E2E5E9;
}

.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 10px;
}

.c5 > * {
  display: block;
  width: 24px;
  height: 24px;
}

.c5 > *:not(:first-child) {
  margin-left: 8px;
}

.c1 {
  box-sizing: border-box;
  cursor: pointer;
  background: #FFFFFF;
  position: relative;
  border-radius: 4px;
  height: 40px;
  border-color: #8A96A8;
  border-width: 1px;
  border-style: solid;
}

.c1.c1[disabled],
.c1.c1[data-read-only] {
  cursor: default;
}

[data-focus-within] .c0,
.c1:focus {
  border-color: #0949AA;
}

.c1:hover {
  border-color: #0A54C2;
}

.c1:focus {
  outline: none;
}

.c1:active {
  outline: none;
}

.c1[data-status='error'] {
  border-color: #DA0B20;
}

.c1[data-status='success'] {
  border-color: #2ABB5B;
}

.c4 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div>
    <div
      class="c0 c1"
      role="button"
      tabindex="0"
    >
      <div
        class="c2 c3"
      >
        <div
          class="c4"
        >
          Селект
        </div>
        <div
          class="c5"
        >
          <div
            aria-hidden="true"
            class="c6"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Select should render small components 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c6 {
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c6 *[fill^='#'] {
  fill: #626F84;
}

.c6.c6[data-disabled] {
  pointer-events: none;
  fill: #A7B0BE;
}

.c6:not([data-disabled]):hover {
  cursor: pointer;
}

.c6:not([data-disabled]):hover *[fill^='#'] {
  fill: #0A54C2;
}

.c3 {
  position: relative;
  padding: 0 12px;
  color: #2B313B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  border: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 14px;
  line-height: 20px;
  font-family: 'VTB Group UI',sans-serif;
  box-sizing: border-box;
  border-radius: 4px;
  height: 100%;
}

[data-status='error'] .c2 {
  border-color: #DA0B20;
}

[data-read-only] .c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #E2E5E9;
  border-color: #E2E5E9;
}

.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 10px;
}

.c5 > * {
  display: block;
  width: 20px;
  height: 20px;
}

.c5 > *:not(:first-child) {
  margin-left: 8px;
}

.c1 {
  box-sizing: border-box;
  cursor: pointer;
  background: #FFFFFF;
  position: relative;
  border-radius: 4px;
  height: 32px;
  border-color: #8A96A8;
  border-width: 1px;
  border-style: solid;
}

.c1.c1[disabled],
.c1.c1[data-read-only] {
  cursor: default;
}

[data-focus-within] .c0,
.c1:focus {
  border-color: #0949AA;
}

.c1:hover {
  border-color: #0A54C2;
}

.c1:focus {
  outline: none;
}

.c1:active {
  outline: none;
}

.c1[data-status='error'] {
  border-color: #DA0B20;
}

.c1[data-status='success'] {
  border-color: #2ABB5B;
}

.c4 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<body>
    <div>
      <div
        class="c0 c1"
        role="button"
        tabindex="0"
      >
        <div
          class="c2 c3"
        >
          <div
            class="c4"
          >
            Селект
          </div>
          <div
            class="c5"
          >
            <div
              aria-hidden="true"
              class="c6"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": .c6 {
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c6 *[fill^='#'] {
  fill: #626F84;
}

.c6.c6[data-disabled] {
  pointer-events: none;
  fill: #A7B0BE;
}

.c6:not([data-disabled]):hover {
  cursor: pointer;
}

.c6:not([data-disabled]):hover *[fill^='#'] {
  fill: #0A54C2;
}

.c3 {
  position: relative;
  padding: 0 12px;
  color: #2B313B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  border: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 14px;
  line-height: 20px;
  font-family: 'VTB Group UI',sans-serif;
  box-sizing: border-box;
  border-radius: 4px;
  height: 100%;
}

[data-status='error'] .c2 {
  border-color: #DA0B20;
}

[data-read-only] .c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #E2E5E9;
  border-color: #E2E5E9;
}

.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 10px;
}

.c5 > * {
  display: block;
  width: 20px;
  height: 20px;
}

.c5 > *:not(:first-child) {
  margin-left: 8px;
}

.c1 {
  box-sizing: border-box;
  cursor: pointer;
  background: #FFFFFF;
  position: relative;
  border-radius: 4px;
  height: 32px;
  border-color: #8A96A8;
  border-width: 1px;
  border-style: solid;
}

.c1.c1[disabled],
.c1.c1[data-read-only] {
  cursor: default;
}

[data-focus-within] .c0,
.c1:focus {
  border-color: #0949AA;
}

.c1:hover {
  border-color: #0A54C2;
}

.c1:focus {
  outline: none;
}

.c1:active {
  outline: none;
}

.c1[data-status='error'] {
  border-color: #DA0B20;
}

.c1[data-status='success'] {
  border-color: #2ABB5B;
}

.c4 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div>
    <div
      class="c0 c1"
      role="button"
      tabindex="0"
    >
      <div
        class="c2 c3"
      >
        <div
          class="c4"
        >
          Селект
        </div>
        <div
          class="c5"
        >
          <div
            aria-hidden="true"
            class="c6"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
